File: edgequake/sdks/kotlin/src/test/kotlin/io/edgequake/sdk/E2ETest.kt
package io.edgequake.sdk

import io.edgequake.sdk.models.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.MethodOrderer.OrderAnnotation

/**
 * E2E tests for the Kotlin SDK against a live EdgeQuake backend.
 *
 * Run: mvn test -Pe2e  OR  mvn test -Dtest.excludedGroups= -Dtest.groups=e2e
 * Requires: backend at http://localhost:8080
 *
 * WHY: Unit tests with mocks give FALSE confidence.
 * These E2E tests verify every path, header, and response shape
 * against the real EdgeQuake API.
 */
@Tag("e2e")
@TestMethodOrder(OrderAnnotation::class)
class E2ETest {

    companion object {
        private lateinit var client: EdgeQuakeClient

        @JvmStatic
        @BeforeAll
        fun setup() {
            val baseUrl = System.getenv("EDGEQUAKE_BASE_URL") ?: "http://localhost:8080"
            // WHY: Default tenant/user UUIDs match migration-created defaults,
            // so conversations/folders tests work without env vars.
            client = EdgeQuakeClient(
                EdgeQuakeConfig(
                    baseUrl = baseUrl,
                    apiKey = System.getenv("EDGEQUAKE_API_KEY"),
                    tenantId = System.getenv("EDGEQUAKE_TENANT_ID") ?: "00000000-0000-0000-0000-000000000002",
                    userId = System.getenv("EDGEQUAKE_USER_ID") ?: "00000000-0000-0000-0000-000000000001",
                    workspaceId = System.getenv("EDGEQUAKE_WORKSPACE_ID"),
                    timeoutSeconds = 15
                )
            )
        }
    }

    // ── 1. Health ────────────────────────────────────────────────────

    @Test @Order(1)
    fun health() {
        val h = client.health.check()
        assertEquals("healthy", h.status)
        assertNotNull(h.version)
        println("Health: ${h.status} v${h.version} storage=${h.storageMode} llm=${h.llmProviderName}")
    }

    // ── 2. Documents ─────────────────────────────────────────────────

    @Test @Order(2)
    fun documentsListAndUpload() {
        val list = client.documents.list()
        val docs = list.documents ?: list.items ?: emptyList()
        println("Documents: ${list.total} total, page=${list.page}")

        val upload = client.documents.uploadText(
            "Kotlin SDK Test ${System.currentTimeMillis()}",
            "This document tests the Kotlin SDK integration. Unique content: ${System.nanoTime()}"
        )
        assertNotNull(upload.documentId, "documentId should not be null")
        println("Uploaded: ${upload.documentId} status=${upload.status}")

        // Allow processing time
        Thread.sleep(2000)

        val doc = client.documents.get(upload.documentId!!)
        assertNotNull(doc.id)
        println("Document: ${doc.title} status=${doc.status}")
    }

    // ── 3. Graph ─────────────────────────────────────────────────────

    @Test @Order(3)
    fun graphGet() {
        val g = client.graph.get()
        assertNotNull(g.nodes)
        assertNotNull(g.edges)
        println("Graph: ${g.nodes?.size} nodes, ${g.edges?.size} edges")
    }

    @Test @Order(4)
    fun graphSearch() {
        val result = client.graph.search("test")
        assertNotNull(result.nodes)
        println("Search: ${result.nodes?.size} nodes found")
    }

    // ── 5. Entities ──────────────────────────────────────────────────

    @Test @Order(5)
    fun entitiesListAndCreateAndDelete() {
        val list = client.entities.list()
        assertNotNull(list.items)
        println("Entities: ${list.total} total")

        // Cleanup first (idempotent)
        val testName = "KOTLIN_SDK_TEST_ENTITY"
        try { client.entities.delete(testName) } catch (_: Exception) {}

        val created = client.entities.create(
            CreateEntityRequest(
                entityName = testName,
                entityType = "TEST",
                description = "Created by Kotlin SDK E2E test",
                sourceId = "kotlin-e2e"
            )
        )
        assertEquals("success", created.status)
        assertNotNull(created.entity)
        println("Created entity: ${created.entity?.entityName}")

        // Verify exists
        val exists = client.entities.exists(testName)
        assertNotNull(exists.entityId)

        // Get detail
        val detail = client.entities.get(testName)
        assertNotNull(detail.entity)
        assertEquals(testName, detail.entity?.entityName)

        // Delete
        val deleted = client.entities.delete(testName)
        assertEquals("success", deleted.status)
        println("Deleted entity: ${deleted.deletedEntityId}")
    }

    // ── 6. Relationships ─────────────────────────────────────────────

    @Test @Order(6)
    fun relationshipsList() {
        val list = client.relationships.list()
        assertNotNull(list.items)
        println("Relationships: ${list.total} total")
    }

    // ── 7. Query ─────────────────────────────────────────────────────

    @Test @Order(7)
    fun queryExecute() {
        val result = client.query.execute("What is EdgeQuake?")
        assertNotNull(result.answer)
        assertTrue(result.answer!!.isNotEmpty())
        println("Query answer (${result.answer!!.length} chars): ${result.answer!!.take(80)}...")
    }

    // ── 8. Chat ──────────────────────────────────────────────────────

    @Test @Order(8)
    fun chatCompletions() {
        try {
            val result = client.chat.completions(
                ChatCompletionRequest(
                    message = "Hello, what is EdgeQuake?"
                )
            )
            assertNotNull(result.content)
            println("Chat: ${result.content?.take(80)}")
        } catch (e: EdgeQuakeException) {
            // Chat may not be fully implemented
            println("Chat: ${e.statusCode} (expected if not implemented)")
        }
    }

    // ── 9. Tenants ───────────────────────────────────────────────────

    @Test @Order(9)
    fun tenantsList() {
        val list = client.tenants.list()
        assertNotNull(list.items)
        assertTrue(list.items!!.isNotEmpty())
        println("Tenants: ${list.items!!.joinToString { "${it.name} (${it.slug})" }}")
    }

    // ── 10. Users ────────────────────────────────────────────────────

    @Test @Order(10)
    fun usersList() {
        val list = client.users.list()
        assertNotNull(list.users)
        println("Users: ${list.users?.size} users")
    }

    // ── 11. API Keys ─────────────────────────────────────────────────

    @Test @Order(11)
    fun apiKeysList() {
        val list = client.apiKeys.list()
        assertNotNull(list.keys)
        println("API Keys: ${list.keys?.size} keys")
    }

    // ── 12. Tasks ────────────────────────────────────────────────────

    @Test @Order(12)
    fun tasksList() {
        val list = client.tasks.list()
        // Response has "tasks" key
        val tasks = list.tasks ?: list.items ?: emptyList()
        println("Tasks: ${tasks.size} total")
    }

    // ── 13. Pipeline ─────────────────────────────────────────────────

    @Test @Order(13)
    fun pipelineStatus() {
        val status = client.pipeline.status()
        assertNotNull(status.isBusy)
        println("Pipeline: busy=${status.isBusy} pending=${status.pendingTasks} completed=${status.completedTasks}")
    }

    @Test @Order(14)
    fun pipelineMetrics() {
        val metrics = client.pipeline.queueMetrics()
        assertNotNull(metrics.pendingCount)
        println("Queue: pending=${metrics.pendingCount} processing=${metrics.processingCount} workers=${metrics.activeWorkers}")
    }

    // ── 15. Models ───────────────────────────────────────────────────

    @Test @Order(15)
    fun modelsCatalog() {
        val catalog = client.models.catalog()
        assertNotNull(catalog.providers)
        assertTrue(catalog.providers!!.isNotEmpty())
        println("Models: ${catalog.providers!!.joinToString { "${it.name} (${it.models?.size} models)" }}")
    }

    @Test @Order(16)
    fun modelsProviderHealth() {
        val health = client.models.health()
        assertTrue(health.isNotEmpty())
        val first = health.first()
        assertNotNull(first.name)
        println("Provider Health: ${health.size} providers, first=${first.name} enabled=${first.enabled}")
    }

    @Test @Order(17)
    fun providerStatus() {
        val status = client.models.providerStatus()
        assertNotNull(status.provider)
        println("Provider: provider=${status.provider} embedding=${status.embedding} storage=${status.storage}")
    }

    // ── 18. Conversations (requires tenant/user) ─────────────────────

    @Test @Order(18)
    fun conversationsCRUD() {
        val conv = client.conversations.create("Kotlin E2E Test ${System.currentTimeMillis()}")
        assertNotNull(conv.id)
        println("Created conversation: ${conv.id} title=${conv.title}")

        val convos = client.conversations.list()
        assertTrue(convos.isNotEmpty())
        println("Conversations: ${convos.size}")

        val detail = client.conversations.get(conv.id!!)
        assertNotNull(detail.conversation)
        assertEquals(conv.id, detail.conversation?.id)

        client.conversations.delete(conv.id!!)
    }

    // ── 19. Folders ──────────────────────────────────────────────────

    @Test @Order(19)
    fun foldersCRUD() {
        val folder = client.folders.create("Kotlin E2E Folder ${System.currentTimeMillis()}")
        assertNotNull(folder.id)
        println("Created folder: ${folder.id} name=${folder.name}")

        val folders = client.folders.list()
        assertTrue(folders.isNotEmpty())
        println("Folders: ${folders.size}")

        client.folders.delete(folder.id!!)
    }

    // ── 20. Costs ────────────────────────────────────────────────────

    @Test @Order(20)
    fun costsSummary() {
        val costs = client.costs.summary()
        assertNotNull(costs.totalCost)
        println("Costs: total=${costs.totalCost} documents=${costs.documentCount} queries=${costs.queryCount}")
    }
}

File: edgequake/sdks/go/e2e_test.go
//go:build e2e

package edgequake_test

import (
	"context"
	"os"
	"testing"
	"time"

	edgequake "github.com/edgequake/edgequake-go"
)

func e2eBaseURL() string {
	if u := os.Getenv("EDGEQUAKE_BASE_URL"); u != "" {
		return u
	}
	return "http://localhost:8080"
}

func e2eClient() *edgequake.Client {
	opts := []edgequake.Option{
		edgequake.WithBaseURL(e2eBaseURL()),
		edgequake.WithTimeout(15 * time.Second),
	}
	if key := os.Getenv("EDGEQUAKE_API_KEY"); key != "" {
		opts = append(opts, edgequake.WithAPIKey(key))
	}
	// WHY: Default migration-created tenant/user always available for E2E
	tid := os.Getenv("EDGEQUAKE_TENANT_ID")
	if tid == "" {
		tid = "00000000-0000-0000-0000-000000000002"
	}
	opts = append(opts, edgequake.WithTenantID(tid))

	uid := os.Getenv("EDGEQUAKE_USER_ID")
	if uid == "" {
		uid = "00000000-0000-0000-0000-000000000001"
	}
	opts = append(opts, edgequake.WithUserID(uid))

	if wid := os.Getenv("EDGEQUAKE_WORKSPACE_ID"); wid != "" {
		opts = append(opts, edgequake.WithWorkspaceID(wid))
	}
	return edgequake.NewClient(opts...)
}

func TestE2E_Health(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	h, err := c.Health.Check(ctx)
	if err != nil {
		t.Fatalf("Health check failed: %v", err)
	}
	if h.Status != "healthy" {
		t.Fatalf("expected healthy, got %s", h.Status)
	}
	if h.Version == "" {
		t.Fatal("version is empty")
	}
	t.Logf("Health: status=%s version=%s storage=%s llm=%s",
		h.Status, h.Version, h.StorageMode, h.LLMProvider)
}

func TestE2E_Documents_ListAndUpload(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()

	list, err := c.Documents.List(ctx, 1, 10)
	if err != nil {
		t.Fatalf("Documents.List failed: %v", err)
	}
	t.Logf("Documents: %d total", len(list.Documents))

	body := map[string]interface{}{
		"content": "EdgeQuake is an advanced Retrieval-Augmented Generation RAG framework implemented in Rust. " +
			"It uses graph-based knowledge representation for enhanced retrieval and entity extraction " +
			"with deduplication algorithms and multi-provider LLM support for production deployments.",
		"title": "E2E Test Document",
	}
	upload, err := c.Documents.UploadText(ctx, body)
	if err != nil {
		t.Fatalf("Documents.UploadText failed: %v", err)
	}
	if upload.ID == "" {
		t.Fatal("upload response has no document_id")
	}
	t.Logf("Uploaded document: id=%s status=%s", upload.ID, upload.Status)

	doc, err := c.Documents.Get(ctx, upload.ID)
	if err != nil {
		t.Fatalf("Documents.Get(%s) failed: %v", upload.ID, err)
	}
	t.Logf("Document: id=%s title=%s status=%s", doc.ID, doc.Title, doc.Status)

	err = c.Documents.Delete(ctx, upload.ID)
	if err != nil {
		t.Logf("Warning: Documents.Delete(%s) failed: %v", upload.ID, err)
	}
}

func TestE2E_Graph_Get(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	g, err := c.Graph.Get(ctx, 10)
	if err != nil {
		t.Fatalf("Graph.Get failed: %v", err)
	}
	t.Logf("Graph: %d nodes, %d edges", len(g.Nodes), len(g.Edges))
}

func TestE2E_Graph_Search(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	results, err := c.Graph.Search(ctx, "test", 5)
	if err != nil {
		t.Fatalf("Graph.Search failed: %v", err)
	}
	t.Logf("Search results: %d nodes", len(results.Nodes))
}

func TestE2E_Entities_ListAndCreate(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()

	// WHY: Clean up any leftover entity from previous runs to make test idempotent.
	_ = c.Entities.Delete(ctx, "E2E_TEST_ENTITY")

	resp, err := c.Entities.List(ctx, 1, 10, "")
	if err != nil {
		t.Fatalf("Entities.List failed: %v", err)
	}
	t.Logf("Entities: %d total (page %d)", resp.Total, resp.Page)

	created, err := c.Entities.Create(ctx, &edgequake.CreateEntityParams{
		EntityName:  "E2E_TEST_ENTITY",
		EntityType:  "TEST",
		Description: "Created by Go SDK E2E test",
		SourceID:    "manual_entry",
	})
	if err != nil {
		t.Fatalf("Entities.Create failed: %v", err)
	}
	t.Logf("Created entity: %s status=%s", created.Entity.EntityName, created.Status)

	exists, err := c.Entities.Exists(ctx, "E2E_TEST_ENTITY")
	if err != nil {
		t.Fatalf("Entities.Exists failed: %v", err)
	}
	if !exists.Exists {
		t.Fatal("entity should exist after creation")
	}

	got, err := c.Entities.Get(ctx, "E2E_TEST_ENTITY")
	if err != nil {
		t.Fatalf("Entities.Get failed: %v", err)
	}
	if got.Entity == nil || got.Entity.EntityName != "E2E_TEST_ENTITY" {
		name := ""
		if got.Entity != nil {
			name = got.Entity.EntityName
		}
		t.Fatalf("got name %s", name)
	}

	err = c.Entities.Delete(ctx, "E2E_TEST_ENTITY")
	if err != nil {
		t.Logf("Warning: Entities.Delete failed: %v", err)
	}
}

func TestE2E_Relationships_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	resp, err := c.Relationships.List(ctx, 1, 10)
	if err != nil {
		t.Fatalf("Relationships.List failed: %v", err)
	}
	t.Logf("Relationships: %d total (page %d)", resp.Total, resp.Page)
}

func TestE2E_Query_Execute(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	resp, err := c.Query.Execute(ctx, &edgequake.QueryRequest{
		Query: "What is EdgeQuake?",
		Mode:  "hybrid",
	})
	if err != nil {
		t.Logf("Query.Execute returned error (may be expected): %v", err)
		return
	}
	t.Logf("Query response: answer=%q sources=%d",
		truncStr(resp.Answer, 80), len(resp.Sources))
}

func TestE2E_Chat_Completions(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	// WHY: EdgeQuake chat API uses `message` (singular string), not messages array
	resp, err := c.Chat.Completions(ctx, &edgequake.ChatCompletionRequest{
		Message: "Hello, what is EdgeQuake?",
	})
	if err != nil {
		t.Logf("Chat.Completions returned error (may need LLM): %v", err)
		return
	}
	if resp.Content != "" {
		t.Logf("Chat response: %q", truncStr(resp.Content, 80))
	}
	if resp.ConversationID != "" {
		t.Logf("Conversation ID: %s", resp.ConversationID)
	}
}

func TestE2E_Conversations_CRUD(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()

	conv, err := c.Conversations.Create(ctx, &edgequake.CreateConversationParams{
		Title: "Go SDK E2E Test Conversation",
	})
	if err != nil {
		t.Fatalf("Conversations.Create failed: %v", err)
	}
	t.Logf("Created conversation: id=%s title=%s", conv.ID, conv.Title)

	convs, err := c.Conversations.List(ctx)
	if err != nil {
		t.Fatalf("Conversations.List failed: %v", err)
	}
	t.Logf("Conversations: %d total", len(convs))

	detail, err := c.Conversations.Get(ctx, conv.ID)
	if err != nil {
		t.Fatalf("Conversations.Get failed: %v", err)
	}
	t.Logf("Conversation detail: id=%s messages=%d",
		detail.ID, len(detail.Messages))

	msg, err := c.Conversations.CreateMessage(ctx, conv.ID,
		&edgequake.CreateMessageParams{
			Role:    "user",
			Content: "Hello from Go SDK E2E test",
		})
	if err != nil {
		t.Fatalf("Conversations.CreateMessage failed: %v", err)
	}
	t.Logf("Created message: id=%s role=%s", msg.ID, msg.Role)

	err = c.Conversations.Delete(ctx, conv.ID)
	if err != nil {
		t.Logf("Warning: Conversations.Delete failed: %v", err)
	}
}

func TestE2E_Folders_CRUD(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()

	folder, err := c.Folders.Create(ctx, &edgequake.CreateFolderParams{
		Name: "Go SDK E2E Test Folder",
	})
	if err != nil {
		t.Fatalf("Folders.Create failed: %v", err)
	}
	t.Logf("Created folder: id=%s name=%s", folder.ID, folder.Name)

	folders, err := c.Folders.List(ctx)
	if err != nil {
		t.Fatalf("Folders.List failed: %v", err)
	}
	t.Logf("Folders: %d total", len(folders))

	err = c.Folders.Delete(ctx, folder.ID)
	if err != nil {
		t.Logf("Warning: Folders.Delete failed: %v", err)
	}
}

func TestE2E_Tasks_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	tasks, err := c.Tasks.List(ctx, "", 1, 10)
	if err != nil {
		t.Fatalf("Tasks.List failed: %v", err)
	}
	t.Logf("Tasks: %d total", tasks.Total)
}

func TestE2E_Pipeline_Status(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	status, err := c.Pipeline.Status(ctx)
	if err != nil {
		t.Fatalf("Pipeline.Status failed: %v", err)
	}
	t.Logf("Pipeline: status=%s active=%d queued=%d",
		status.Status, status.ActiveTasks, status.QueuedTasks)
}

func TestE2E_Pipeline_Metrics(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	metrics, err := c.Pipeline.Metrics(ctx)
	if err != nil {
		t.Fatalf("Pipeline.Metrics failed: %v", err)
	}
	t.Logf("Queue metrics: depth=%d processing=%d",
		metrics.QueueDepth, metrics.Processing)
}

func TestE2E_Costs_Summary(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	summary, err := c.Costs.Summary(ctx)
	if err != nil {
		t.Fatalf("Costs.Summary failed: %v", err)
	}
	t.Logf("Costs: total=$%.4f tokens=%d",
		summary.TotalCostUSD, summary.TotalTokens)
}

func TestE2E_Costs_Budget(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	budget, err := c.Costs.Budget(ctx)
	if err != nil {
		t.Fatalf("Costs.Budget failed: %v", err)
	}
	t.Logf("Budget: current_spend=$%.4f", budget.CurrentSpendUSD)
}

func TestE2E_Models_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	catalog, err := c.Models.List(ctx)
	if err != nil {
		t.Fatalf("Models.List failed: %v", err)
	}
	t.Logf("Models: %d providers", len(catalog.Providers))
	for _, p := range catalog.Providers {
		t.Logf("  - %s (%s): %d models",
			p.Name, p.DisplayName, len(p.Models))
	}
}

func TestE2E_Models_ProviderStatus(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	status, err := c.Models.ProviderStatus(ctx)
	if err != nil {
		t.Fatalf("Models.ProviderStatus failed: %v", err)
	}
	t.Logf("Provider: %s model=%s status=%s",
		status.CurrentProvider, status.CurrentModel, status.Status)
}

func TestE2E_Models_ProviderHealth(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	providers, err := c.Models.ProviderHealth(ctx)
	if err != nil {
		t.Fatalf("Models.ProviderHealth failed: %v", err)
	}
	t.Logf("Provider health: %d providers", len(providers))
}

func TestE2E_Tenants_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	resp, err := c.Tenants.List(ctx)
	if err != nil {
		t.Fatalf("Tenants.List failed: %v", err)
	}
	t.Logf("Tenants: %d total", resp.Total)
	for _, ten := range resp.Items {
		t.Logf("  - %s (slug=%s plan=%s)", ten.Name, ten.Slug, ten.Plan)
	}
}

func TestE2E_Users_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	resp, err := c.Users.List(ctx)
	if err != nil {
		t.Fatalf("Users.List failed: %v", err)
	}
	t.Logf("Users: %d total", resp.Total)
}

func TestE2E_APIKeys_List(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	resp, err := c.APIKeys.List(ctx)
	if err != nil {
		t.Fatalf("APIKeys.List failed: %v", err)
	}
	t.Logf("API Keys: %d total", resp.Total)
}

func TestE2E_Lineage_ForEntity(t *testing.T) {
	c := e2eClient()
	ctx := context.Background()
	graph, err := c.Lineage.ForEntity(ctx, "EDGEQUAKE", 2)
	if err != nil {
		t.Logf("Lineage.ForEntity returned error (expected): %v", err)
		return
	}
	t.Logf("Lineage: %d nodes, %d edges", len(graph.Nodes), len(graph.Edges))
}

func truncStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
